text/0000-read-buf.md

```diff
@@ -0,0 +1,655 @@
+                return Err(io::Error::new(io::ErrorKind::UnexpectedEof, "failed to fill buffer"));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        let initial_len = buf.len();
+
+        let mut initialized = 0;
+        loop {
+            if buf.len() == buf.capacity() {
+                buf.reserve(32);
+            }
+
+            let mut read_buf = ReadBuf::uninit(buf.spare_capacity_mut());
+            unsafe {
+                read_buf.assume_init(initialized);
+            }
+
+            match self.read_buf(&mut read_buf) {
+                Ok(()) => {}
+                Err(e) if e.kind() = io::ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e),
+            }
+
+            if read_buf.filled().is_empty() {
+                break;
+            }
+
+            initialized = read_buf.initialized().len() - read_buf.filled().len();
+            let new_len = buf.len() + read_buf.filled().len();
+            unsafe {
+                buf.set_len(new_len);
+            }
+        }
+
+        Ok(buf.len() - initial_len)
+    }
+}
+
+pub fn copy<R, W>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+where
+    R: Read,
+    W: Write,
+{
+    let mut buf = [MaybeUninit::uninit(); 4096];
+    let mut buf = ReadBuf::uninit(&mut buf);
+    let mut len = 0;
+
+    loop {
+        match reader.read_buf(&mut buf) {
+            Ok(()) => {},
+            Err(e) if e.kind() == io::ErrorKind::Interrupted => continue,
+            Err(e) => return Err(e),
+        };
+
+        if buf.filled().is_empty() {
+            break;
+        }
+
+        len += buf.filled().len() as u64;
+        writer.write_all(buf.filled())?;
+        buf.clear();
+    }
+
+    Ok(len)
+}
+```
+
+The existing `std::io::Initializer` type and `Read::initializer` method will be removed.
+
+Vectored reads use a similar API:
+
+```rust
+/// A possibly-uninitialized version of `IoSliceMut`.
+///
+/// It is guaranteed to have exactly the same layout and ABI as `IoSliceMut`.
+pub struct MaybeUninitIoSliceMut<'a> { ... }
+
+impl<'a> MaybeUninitIoSliceMut<'a> {
+    /// Creates a new `MaybeUninitIoSliceMut` from a slice of maybe-uninitialized bytes.
+    #[inline]
+    pub fn new(buf: &'a mut [MaybeUninit<u8>]) -> MaybeUninitIoSliceMut<'a> { ... }
+}
+
+impl<'a> Deref for MaybeUninitIoSliceMut<'a> {
+    type Target = [MaybeUninit<u8>];
+
+    ...
+}
+
+impl<'a> DerefMut for MaybeUninitIoSliceMut<'a> { ... }
+
+
+/// A wrapper over a set of incrementally-initialized buffers.
+pub struct ReadBufs<'a> { ... }
+
+impl<'a> ReadBufs<'a> {
+    /// Creates a new `ReadBufs` from a set of fully initialized buffers.
+    #[inline]
+    pub fn new(bufs: &'a mut [IoSliceMut<'a>]) -> ReadBufs<'a> { ... }
+
+    /// Creates a new `ReadBufs` from a set of fully uninitialized buffers.
+    ///
+    /// Use `assume_init` if part of the buffers are known to be already initialized.
```

Comment:
`assume_init` isn't defined on the `ReadBufs` type?

I think `ReadBufs` could do with some elaboration: for example, why is it not just a slice of `ReadBuf` types? What does its API surface look like beyond `new` and `uninit`?

Reply:
Also, there should probably be a way to convert a `ReadBuf` into a `ReadBufs` so that `read_buf` can easily delegate to `read_buf_vectored`.